lib/src/base.dart

```diff
@@ -7,3 +7,17 @@ import 'visitors.dart';
 abstract class Spec {
   R accept<R>(SpecVisitor<R> visitor, [R context]);
 }
+
+/// Returns a generic [Spec] that is lazily generated when visited.
+Spec lazySpec(Spec Function() generate) => new _LazySpec(generate);
+
+class _LazySpec implements Spec {
+  final Spec Function() generate;
+
+  const _LazySpec(this.generate);
+
+  @override
+  R accept<R>(SpecVisitor<R> visitor, [R context]) {
+    return generate().accept(visitor, context);
```

Comment:
This not only generates lazily, but also generates _every time_ it's visited. Is it intentional?

Reply:
Yeah. We could cache it, but we wouldn't know if the function is stateless or not.